File: Problems/0946. (M) Validate Stack Sequences/Python3/solution.py
class Solution:
  def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
    # stack with two pointers

    pop_ptr = 0
    stack = []
    for i in range(len(pushed)):
      stack.append(pushed[i])
      # keep matching until mismatch detected
      while stack and pop_ptr < len(popped) and stack[-1] == popped[pop_ptr]:
        stack.pop()
        pop_ptr += 1
    
    return len(popped) == pop_ptr

File: Problems/0946. (M) Validate Stack Sequences/Java/solution.java
class Solution {
  public boolean validateStackSequences(int[] pushed, int[] popped) {
    // stack with two pointers

    int pop_ptr = 0;
    Deque<Integer> stack = new ArrayDeque();
    for(int i = 0; i < pushed.length; i++) {
      stack.push(pushed[i]);
      // keep matching until mismatch detected
      while (stack.size() > 0 && pop_ptr < popped.length && stack.peek() == popped[pop_ptr]) {
        stack.pop();
        pop_ptr++;
      }
    }
    return popped.length == pop_ptr;
  }
}
